### PR TITLE
Add DCGM data source and extractor for GPU utilization

### PR DESCRIPTION
**Pull request by noalimoy**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

The EPP routes requests using application-level metrics (KV cache, queue
depth) but has no GPU hardware utilization signal. This adds the
data-layer infrastructure to collect GPU metrics from a DCGM Exporter endpoints:

- `HTTPDataSource` port override to scrape a different port on the same
  pod IP (backward-compatible, default 0 preserves existing behavior)
- `dcgm-data-source` plugin that polls and parses Prometheus metrics
- `dcgm-extractor` plugin that reads `DCGM_FI_DEV_GPU_UTIL`, aggregates
  across GPUs (max), and stores normalized utilization in the endpoint's
  `AttributeMap`
- `GPUUtilization` attribute type with `Produces`/`Consumes` data key

A follow-up PR adds a filter and scorer that consume this attribute.

**Which issue(s) this PR fixes**:
Part of #1706

**Release note** _(write `NONE` if no user-facing change)_:
```release-note
NONE
```

### CODE REVIEW

File: pkg/epp/framework/plugins/datalayer/attribute/gpu/data_types.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// GPUUtilization is a normalized GPU compute utilization value in [0.0, 1.0],
+// derived from DCGM_FI_DEV_GPU_UTIL (which reports 0-100).
+// For multi-GPU pods the extractor aggregates across visible devices.
+type GPUUtilization float64
+
+func (v GPUUtilization) Clone() fwkdl.Cloneable {
+	return v
+}
+
+// ReadGPUUtilization retrieves GPU utilization from an endpoint's AttributeMap.
```
Comment (liu-cong) [inline pkg/epp/framework/plugins/datalayer/attribute/gpu/data_types.go]: Nit: callers always pass `GPUUtilizationDataKey.WithNonEmptyProducerName(...).String()`. Consider making this a no-arg helper that encapsulates the key construction, since there's only one producer. Not blocking.

File: pkg/epp/framework/plugins/datalayer/source/http/datasource.go [hunk 2]
```diff
@@ -61,7 +68,9 @@ type HTTPDataSource[T any] struct {
 }
 
 // NewHTTPDataSource constructs a typed polling dispatcher.
-func NewHTTPDataSource[T any](scheme, path string, skipCertVerification bool,
+// portOverride, when non-zero, makes the source scrape podIP:portOverride
+// instead of the endpoint's MetricsHost.
+func NewHTTPDataSource[T any](scheme, path string, portOverride int, skipCertVerification bool,
```
Comment (liu-cong) [inline pkg/epp/framework/plugins/datalayer/source/http/datasource.go]: Can we create an `Option` struct for optional args such as the port override, so `NewHTTPDataSource(existingargs, opts... Option)`

Comment (noalimoy) [reply pkg/epp/framework/plugins/datalayer/source/http/datasource.go]: Fixed.

File: pkg/epp/framework/plugins/datalayer/source/dcgm/datasource.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+// NewHTTPDCGMDataSource constructs a DCGMDataSource with the given parameters.
+// Use this function directly in tests to bypass JSON parameter marshaling.
+func NewHTTPDCGMDataSource(scheme, path string, port int, name string) (*http.HTTPDataSource[metrics.PrometheusMetricMap], error) {
+	return http.NewHTTPDataSource(scheme, path, defaultDCGMInsecureSkipVerify,
+		DCGMDataSourceType, name, parsePrometheus,
+		http.WithPortOverride(port))
+}
+
+// DCGMDataSourceFactory instantiates a dcgm-data-source plugin from configuration.
+func DCGMDataSourceFactory(name string, parameters *json.Decoder, _ plugin.Handle) (plugin.Plugin, error) {
```
Comment (elevran) [inline pkg/epp/framework/plugins/datalayer/source/dcgm/datasource.go:59]: should-fix: this package has no test files. DCGMDataSourceFactory's default params and scheme validation aren't covered - worth mirroring source/metrics/datasource_config_test.go's pattern for the defaults and invalid-scheme case.

**Comment by elevran** [comment]
@noalimoy PR has been review and needs rebase/conflict resolution to progress.

**Review by elevran — COMMENTED** [review]
Clean, well-tested addition - the extractor logic (missing metric, empty samples, multi-GPU max aggregation, normalization) is covered thoroughly, and the port-override mechanism on HTTPDataSource is backward-compatible for existing callers (variadic option, no signature break for source/metrics or source/models).

**Comment by noalimoy** [comment]
Thanks @elevran for the review!
I've been out sick since last Tuesday (fell on my way to work:(  ) - and just got back today. I already have a local patch ready from last week for the DaemonSet support (`useNodeAddress` + pod-label filtering in the extractor), will push it soon along with `source/dcgm` factory tests.

File: pkg/epp/framework/plugins/datalayer/source/http/datasource.go [hunk 4]
```diff
@@ -231,5 +272,15 @@ func (s *HTTPDataSource[T]) AppendExtractor(ext fwkplugin.Plugin) error {
 }
 
 func (s *HTTPDataSource[T]) getEndpoint(ep Addressable) *url.URL {
-	return &url.URL{Scheme: s.scheme, Host: ep.GetMetricsHost(), Path: s.path}
+	host := ep.GetMetricsHost()
+	if s.portOverride > 0 {
```
Comment (elevran) [inline pkg/epp/framework/plugins/datalayer/source/http/datasource.go:279]: `useNodeAddress` is only honored inside the `portOverride > 0` branch. 
Consider validating `useNodeAddress` requires a non-zero `portOverride` at construction time.

File: pkg/epp/framework/plugins/datalayer/extractor/dcgm/extractor.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+			continue
+		}
+		val := gaugeValue(m)
+		if val > maxUtil {
+			maxUtil = val
+		}
+		found = true
+	}
+	if !found {
+		return fmt.Errorf("metric %q present but has no samples for pod %q", gpuUtilMetricName, podName)
```
Comment (elevran) [inline pkg/epp/framework/plugins/datalayer/extractor/dcgm/extractor.go:101]: Both this and the "metric not found" branch above return hard errors, incrementing the `DataLayerExtractErrorsTotal` on every poll cycle. 
For DaemonSet mode, this fires continuously whenever a pod isn't yet visible to the exporter (exporter restart/lag, pod just scheduled) - not necessarily a real error each time. 
While this matches and is consistent with the existing `models`/`metrics` extractor convention, which also hard-errors on a miss, worth confirming the resulting error-counter noise is acceptable for DaemonSet steady state with slow-starting pods.

File: pkg/epp/framework/plugins/datalayer/source/dcgm/datasource.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/llm-d/llm-d-router/pkg/epp/framework/plugins/datalayer/source/metrics"
+)
+
+const DCGMDataSourceType = "dcgm-data-source"
+
+const (
+	defaultDCGMScheme             = "http"
+	defaultDCGMPath               = "/metrics"
+	defaultDCGMPort               = 9400
+	defaultDCGMInsecureSkipVerify = true
```
Comment (elevran) [inline pkg/epp/framework/plugins/datalayer/source/dcgm/datasource.go:38]: `insecureSkipVerify` defaults to `true` and the default `scheme` is `http` anyway, making the TLS default moot in the common case. This matches the other HTTP data sources so just flagging.
No action needed unless the DCGM uses TLS by default.

File: pkg/epp/framework/plugins/datalayer/source/dcgm/datasource.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+	Path               string `json:"path"`
+	Port               int    `json:"port"`
+	InsecureSkipVerify bool   `json:"insecureSkipVerify"`
+	// UseNodeAddress scrapes the node IP (DaemonSet) instead of the pod IP (sidecar).
+	UseNodeAddress bool `json:"useNodeAddress"`
+}
+
+// NewHTTPDCGMDataSource constructs a DCGMDataSource with the given parameters.
+// Use this function directly in tests to bypass JSON parameter marshaling.
+func NewHTTPDCGMDataSource(scheme, path string, port int, name string, opts ...http.Option) (*http.HTTPDataSource[metrics.PrometheusMetricMap], error) {
```
Comment (elevran) [inline pkg/epp/framework/plugins/datalayer/source/dcgm/datasource.go:52]: nit: `NewHTTPDCGMDataSource` isn't called anywhere in this PR (including its own tests). Expect it's waiting on the addition of DCGM based filter/scorer.
May want to cover in tests in the meantime, otherwise it's unused exported symbol.

File: pkg/epp/framework/plugins/datalayer/extractor/dcgm/extractor.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+// Produces declares the data key this extractor publishes.
+func (e *Extractor) Produces() map[fwkplugin.DataKey]any {
+	return map[fwkplugin.DataKey]any{e.dk: attrgpu.GPUUtilization(0)}
+}
+
+// sampleBelongsToPod reports whether m should contribute to this endpoint's
+// utilization. When podName is empty, or the sample has no pod label, the
+// sample is kept (sidecar / single-tenant payloads). When both are set, only
+// matching samples are kept (DaemonSet multi-pod payloads).
+func sampleBelongsToPod(m *dto.Metric, podName string) bool {
```
Comment (elevran) [inline pkg/epp/framework/plugins/datalayer/extractor/dcgm/extractor.go:118]: Unlabeled samples are kept regardless of `podName` (L121). Documented and intentional for the sidecar case, but for DaemonSet mode, if the exporter ever emits a mix of labeled and unlabeled samples for the same metric family (e.g., partial misconfiguration, mixed exporter versions during a rollout), unlabeled samples get silently folded into every pod's `max` aggregation. 
Edge case, not a bug given the explicit comment/tests - just flagging to confirm this is desired behavior.

File: pkg/epp/framework/plugins/datalayer/extractor/dcgm/extractor.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+	for _, l := range m.GetLabel() {
+		if l.GetName() == name {
+			return l.GetValue(), true
+		}
+	}
+	return "", false
+}
+
+func gaugeValue(m *dto.Metric) float64 {
+	if g := m.GetGauge(); g != nil {
```
Comment (elevran) [inline pkg/epp/framework/plugins/datalayer/extractor/dcgm/extractor.go:139]: nit: returns `0` for a non-gauge metric rather than erroring, so if `DCGM_FI_DEV_GPU_UTIL` were ever exposed as a non-gauge type this would silently report 0% utilization instead of failing loudly. Low risk of DCGM changes, hence might want to err on the more strict side to surface.

**Review by elevran — COMMENTED** [review]
PR is well scoped and follows repo's conventions. 
A few points below worth a look before merge; none are hard blockers.

File: pkg/epp/framework/plugins/datalayer/source/http/datasource.go (hunk 4 above)
Comment (noalimoy) [reply pkg/epp/framework/plugins/datalayer/source/http/datasource.go:279]: added construction-time validation: `WithUseNodeAddress` now returns an error if `portOverride` is zero.

File: pkg/epp/framework/plugins/datalayer/extractor/dcgm/extractor.go (hunk 5 above)
Comment (noalimoy) [reply pkg/epp/framework/plugins/datalayer/extractor/dcgm/extractor.go:101]: Yes, you're right - matches existing extractor convention. Noise is bounded to pod churn and self-heals once the exporter sees the pod.

File: pkg/epp/framework/plugins/datalayer/source/dcgm/datasource.go (hunk 7 above)
Comment (noalimoy) [reply pkg/epp/framework/plugins/datalayer/source/dcgm/datasource.go:52]: added direct tests for `NewHTTPDCGMDataSource` (with and without `WithUseNodeAddress`).

File: pkg/epp/framework/plugins/datalayer/extractor/dcgm/extractor.go (hunk 8 above)
Comment (noalimoy) [reply pkg/epp/framework/plugins/datalayer/extractor/dcgm/extractor.go:118]: intentional. DaemonSet mode requires `DCGM_EXPORTER_KUBERNETES=true` which labels all samples. Mixed labeled/unlabeled is a deployment misconfiguration, not a runtime scenario we need to guard against in code.

File: pkg/epp/framework/plugins/datalayer/extractor/dcgm/extractor.go (hunk 9 above)
Comment (noalimoy) [reply pkg/epp/framework/plugins/datalayer/extractor/dcgm/extractor.go:139]: made `gaugeValue` strict - returns an error for non-gauge metrics instead of silent 0.

**Review by elevran — DISMISSED** [review]

**Review by elevran — APPROVED** [review]
